### PR TITLE
Allow specifiying inputs as GradientEdge in autograd APIs

### PR DESCRIPTION
**Pull request by albanD**
This can be useful for advanced users (like AOTAutograd) who don't want to keep the corresponding Tensor alive (for memory reasons for example) or when inplace op will change the Tensor's grad_fn (but gradients wrt to the original value is needed).

I went minimal API change but open to suggestions.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: test/test_autograd.py [hunk 1]
```diff
@@ -798,6 +800,73 @@ def test_grad(self):
...
+            ref = torch.autograd.grad(out.sum(), inp)
+
+            grad_fn = inp.grad_fn if inp.grad_fn else inp.clone().grad_fn.next_functions[0][0]
+            edge = torch.autograd.graph.GradientEdge(grad_fn, inp.output_nr)
+
+            new = torch.autograd.grad(out.sum(), edge)
+            self.assertEqual(ref, new)
+
+        # We need to ensure that our main types of Node work (regular cpp Nodes,
+        # AccumulateGrad Nodes and custom Function)
```
Comment (soulitzer) [inline test/test_autograd.py:814]: Maybe add testing for the multiple output case, e.g. with var_mean?
The inplace case might also be interesting to test.

File: torch/autograd/graph.py [hunk 2]
```diff
@@ -133,6 +134,14 @@ def __subclasshook__(cls, C):
         return NotImplemented
 
 
+GradientEdge = namedtuple("GradientEdge", ("node output_nr"))
+GradientEdge.__doc__ = """\
+Object representing a given gradient edge within the autograd graph.
+To get the gradient edge where a given Tensor gradient will be computed,
+you can do ``edge = GradientEdge(t.grad_fn, t.output_nr)``.
+"""
```
Comment (soulitzer) [inline torch/autograd/graph.py:142]: Do we want to mention how its used here? Like computing gradient wrt `t` is equivalent to computing gradient wrt to its gradient edge.

**Review by soulitzer — APPROVED** [review]
Super cool!!

File: torch/csrc/autograd/python_engine.cpp [hunk 3]
```diff
@@ -289,41 +290,62 @@ PyObject* THPEngine_run_backward(
...
+          node_sp = ((torch::autograd::THPCppFunction*)node)->cdata;
+        }
+
+        auto output_nr = THPUtils_unpackUInt32(PyTuple_GetItem(input, 1));
+        output_edges.emplace_back(node_sp, output_nr);
       } else {
-        output_edges.emplace_back(grad_fn, output_nr);
+        THPUtils_assert(
+            false,
+            "all inputs have to be Tensors or GradiendEdges, but got %s",
```
Comment (soulitzer) [inline torch/csrc/autograd/python_engine.cpp:347]:
```suggestion
            "all inputs have to be Tensors or GradientEdges, but got %s",
```

File: torch/autograd/graph.py (hunk 2 above)
Comment (soulitzer) [reply torch/autograd/graph.py:142]: It would seem convenient to have a helper API that allows users to create a GradientEdge from a tensor `edge = get_gradient_edge(x)`

Comment (albanD) [reply torch/autograd/graph.py:142]: And we can do the fancy trick to get the accumulate grad node in there? I like that

**Comment by ezyang** [comment]
Thanks Alban!

File: torch/autograd/graph.py [hunk 4]
```diff
@@ -133,6 +134,29 @@ def __subclasshook__(cls, C):
...
+    if not tensor.requires_grad:
+        raise RuntimeError("It is not possible to get the gradient edge for a Tensor that does not require gradients")
+    grad_fn = tensor.grad_fn
+    if grad_fn is None:
+        # Do an op to force AccumulateGrad lazy creation and get it
+        grad_fn = tensor.view_as(tensor).grad_fn.next_functions[0][0]
+
+    # Note that output_nr default to 0 which is the right value
+    # for the AccumulateGrad node.
+    return GradientEdge(grad_fn, tensor.output_nr)
```
Comment (soulitzer) [inline torch/autograd/graph.py:158]: Don't forgot to update `autograd.rst`'s "Autograd graph" section to generate the docs!

File: torch/autograd/__init__.py [hunk 5]
```diff
@@ -337,16 +337,18 @@ def grad(
         Tuple[torch.Tensor, ...],
         (outputs,) if is_tensor_like(outputs) else tuple(outputs),
     )
-    t_inputs = cast(
-        Tuple[torch.Tensor, ...], (inputs,) if is_tensor_like(inputs) else tuple(inputs)
-    )
+    if is_tensor_like(inputs) or isinstance(inputs, graph.GradientEdge):
+        inputs = cast(_TensorOrTensorsOrGradEdge, (inputs,))
+    else:
+        inputs = tuple(inputs)
```
Comment (soulitzer) [inline torch/autograd/__init__.py:343]: ahh good catch

**Review by soulitzer — APPROVED** [review]
LGTM

File: torch/autograd/graph.py [hunk 6]
```diff
@@ -15,6 +15,7 @@
     "register_multi_grad_hook",
     "allow_mutation_on_saved_tensors",
     "Node",
+    "GradientEdge",
```
Comment (soulitzer) [inline torch/autograd/graph.py:18]: add get_gradient_edge here too
